### PR TITLE
Enable MSAL to talk to Broker via AccountManager without requiring GET_ACCOUNTS permission

### PR DESCRIPTION
**Pull request by rpdome**

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerBaseStrategy.java [hunk 1]
```diff
@@ -79,53 +94,16 @@ Handler getPreferredHandler() {
...
-        );
-
-        requestBundle.putInt(
-                AuthenticationConstants.Broker.CALLER_INFO_UID,
-                Binder.getCallingUid()
+                new Gson().toJson(
+                        mRequestAdapter.brokerRequestFromAcquireTokenParameters(parameters),
+                        BrokerRequest.class)
         );
+        interactiveRequestIntent.putExtra(AuthenticationConstants.Broker.ACCOUNT_NAME, parameters.getLoginHint());
```
Comment (kreedula) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerBaseStrategy.java]: is this new ? why are we passing this seperately, wouldn't this be part of the `BrokerRequest`?

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java [hunk 2]
```diff
@@ -56,397 +48,306 @@
...
-                                .putErrorDescription(errorMsg)
-                );
-
-                throw new ClientException(
-                        ErrorStrings.BROKER_REQUEST_CANCELLED,
-                        errorMsg,
-                        e
-                );
-            }
+        if (!BrokerMsalController.isAccountManagerPermissionsGranted(parameters.getAppContext())) {
```
[the remark this answers is not in the document]
Comment (rpdome) [reply msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: We still do (for pre API 22)

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerBaseStrategy.java (hunk 1 above)
Comment (rpdome) [reply msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerBaseStrategy.java]: Seems like the only place that uses this (apart from logging) is BrokerAccountService - meaning this is not needed.

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java [hunk 3]
```diff
@@ -56,397 +48,305 @@
...
-                throw new ClientException(
-                        ErrorStrings.BROKER_REQUEST_CANCELLED,
-                        errorMsg,
-                        e
-                );
-            }
+        return result;
+    }
+
+    private static BaseException getBaseExceptionFromException(final Exception e) {
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: We already have a method like this in `ExceptionAdapter.java` within common, why not use that?

File: msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java [hunk 4]
```diff
@@ -384,7 +385,7 @@ private void persistCurrentAccount(@Nullable final List<ICacheRecord> cacheRecor
             return;
         }
 
-        String currentAccountJsonString = MsalBrokerResultAdapter.getJsonStringFromICacheRecordList(cacheRecords);
+        String currentAccountJsonString = JsonExtensions.getJsonStringFromICacheRecordList(cacheRecords);
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java]: nit: final

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java [hunk 5]
```diff
@@ -56,397 +48,305 @@
...
             }
 
-            Telemetry.emit(
-                    new BrokerEndEvent()
-                            .putAction(methodName)
-                            .isSuccessful(true)
-            );
+            @Override
+            public String getMethodName() {
+                return ":helloWithAccountManager";
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: This does not match the actual method name, is that intended ?

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java [hunk 6]
```diff
@@ -56,397 +48,305 @@
...
+            }
 
-            return cacheRecords;
-        }
+            @Override
+            public Boolean getResultFromBundle(Bundle bundle) throws BaseException {
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: nit: final

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java [hunk 7]
```diff
@@ -56,397 +48,305 @@
...
+                    public Bundle getRequestBundle(AcquireTokenOperationParameters parameters) {
+                        final Bundle requestBundle = new Bundle();
+                        requestBundle.putString(AuthenticationConstants.Broker.BROKER_ACCOUNT_MANAGER_OPERATION_KEY,
+                                AuthenticationConstants.BrokerAccountManagerOperation.GET_INTENT_FOR_INTERACTIVE_REQUEST);
+                        return requestBundle;
+                    }
+
+                    @Override
+                    public String getMethodName() {
+                        return ":acquireTokenSilentWithAccountManager";
```
Comment (shahzaibj) [inline msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: This is wrong

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java (hunk 5 above)
Comment (rpdome) [reply msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: Yes.

**Comment by shahzaibj** [comment]
@rpdome Left few comments

File: msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java (hunk 3 above)
Comment (rpdome) [reply msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: It's not exactly same. (i.e. this returns an error specific to this class if the exception can't be matched).

Comment (shahzaibj) [reply msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: May be we can add that as a case to the method in common? 

Leaving it here is fine too. I'll let you pick.

Comment (rpdome) [reply msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerAccountManagerStrategy.java]: We can't. This generic error message is specific to account manager.

**Pull request merged**